enhancements/generic-constraints.md

```diff
@@ -0,0 +1,265 @@
+
+#### Compound Constraint
+
+As an operator author, I would like to specify a constraint that has multiple rules and requirements that must be resolved into a single dependent operator. For example, I would like to have a constraint for a dependent operator that belongs to a specific package and provides a specific API/GVK.
+
+### Design Details
+
+#### Generic Constraint Syntax
+
+The constraint syntax is designed to have similarity with the existing constraints that OLM supports which are GVK and package. It is intentional to keep the fundamental syntax with `type` and `value` the same so that it doesn't introduce any dramatic changes to the overall user experiences.
+
+Each constraint is specified using the following syntax:
+
+```json=
+{
+   "type":"olm.constraint",
+   "value":{
+      "evaluator":{
+         "id":"cel"
+      },
+      "rule": "properties.exists(p, p.type == \"certified\")",
+      "message": "require to have certified property",
+      "action":{
+         "id":"require"
+      }
+   }
+}
+```
+
+The constraint `type` is `olm.constraint` and the `value` is information associated with constraint. The `value` struct has 4 fields: `evaluator`, `rule`, `message` and `action`.
+
+The `evaluator` is a struct with `id` field to represent the language library that will be used to evaluate the expression. At the moment, only CEL expression is supported using `cel` as the identifier. Given the possibility of supporting other expression languages in the future, the `evaluator` is constructed as a struct so that additional fields can be added without breaking the current syntax. For example, if we support a hypothetical expression language named `lucky` that has ability to dynamically load a package named `bobby` at runtime, a `package` field can be added to `evaluator` struct to support that ability:
+
+```json=
+"evaluator":{
+    "id":"lucky",
+    "package":"bobby"
+},
+```
+
+The `rule` field is the string that presents a CEL expression that will be evaluated during resolution against . Only CEL expression is supported in the initial release.
+
+The `message` field is an optional field that is accommodating the rule field to surface information regarding what this CEL rule is about. The message will be surfaced in resolution output when the rule evaluates to false.
+
+Finally, the `action` field is a struct with the `id` field to indicate the resolution action that this constraint will behave. For example, for `require` action, there must be at least one candidate that satisfies this constraint. This action can be used to indicate optional constraint in the future adding a new field `optional` such as:
```

Comment:
I don't really understand an "optional constraint". I understand an optional dependency, meaning that an API / package may need to be installed if possible but not a "constraint". Would you mind clarifying what is behind it, giving an example would be great?

Reply:
Outdated now since the action field is no longer in scope in this EP. However, for posterity...

An explicit package dependency is just a type of constraint.
- A required dependency looks like:  "A valid SAT solution MUST include a package that meets the dependency constraint"
- An optional dependency looks like: "A valid SAT solution CAN include a package that meets the dependency constraint and should be favored over a solution that does not"

Generalizing to arbitrary constraints, I think it would translate to:
- A required constraint looks like: "A valid SAT solution MUST include a package that meets the constraint"
- An optional constraint looks like: "A valid SAT solution CAN include a package that meets the constraint and should be favored over a solution that does not"